class.jetpack.php

```diff
@@ -6929,22 +6930,22 @@ public static function build_raw_urls( $url ) {
 	 * Stores and prints out domains to prefetch for page speed optimization.
 	 *
 	 * @param mixed $new_urls
+	 * @param string | $relation_type | resource hint type
 	 */
-	public static function dns_prefetch( $new_urls = null ) {
+	public static function dns_prefetch( $new_urls = null, $relation_type = 'dns-prefetch' ) {
 		static $prefetch_urls = array();
-		if ( empty( $new_urls ) && ! empty( $prefetch_urls ) ) {
-			echo "\r\n";
-			foreach ( $prefetch_urls as $this_prefetch_url ) {
-				printf( "<link rel='dns-prefetch' href='%s'/>\r\n", esc_attr( $this_prefetch_url ) );
-			}
-		} elseif ( ! empty( $new_urls ) ) {
-			if ( ! has_action( 'wp_head', array( __CLASS__, __FUNCTION__ ) ) ) {
-				add_action( 'wp_head', array( __CLASS__, __FUNCTION__ ) );
-			}
-			foreach ( (array) $new_urls as $this_new_url ) {
-				$prefetch_urls[] = strtolower( untrailingslashit( preg_replace( '#^https?://#i', '//', $this_new_url ) ) );
+		if( 'dns-prefetch' == $relation_type ) {
+			if ( empty( $new_urls ) && ! empty( $prefetch_urls ) ) {
```

Comment:
Since this is on a filter, we can't use the lack of $new_urls as a way to check to see if we need to put them. We could do something like `doing_filter` to account for that and keep it all within one function.

I don't know if keeping it all within one function makes the most sense for long-term sustainability or if we're better off separating it out.

Reply:
OK, I _think_ I'm with you. If I'm understanding what you're suggesting:
1 x function to effectively capture the `$urls` (keep this function, pretty much as it is?) + 1 x function to capture the `$relation_type`, as opposed to trying to mash it all in to one "do it all" function?

Reply:
How do you feel about something like:
```
	if ( 'dns-prefetch' == $relation_type ) {
		foreach ( $urls as $key => $url ) {
			if ( strpos( $url, $prefetch_urls) == false ) {
				foreach ( (array) $new_urls as $this_new_hint ) {
					$hints[] = strtolower( untrailingslashit( preg_replace( '#^https?://#i', '//', $this_prefetch_hint ) ) );
				}
			}
		}
		$hints = array_unique( $hints );
	}
```